src/types/misc/CommunityModel.jl

```diff
@@ -0,0 +1,116 @@
+)
+
+"""
+$(TYPEDSIGNATURES)
+
+Shallow copy of a [`CommunityModel`](@ref)
+"""
+Base.copy(m::CommunityMember) = CommunityMember(;
+    id = m.id,
+    model = m.id,
+    exchange_reaction_ids = m.exchange_reaction_ids,
+    biomass_reaction_id = m.biomass_reaction_id,
+)
+
+"""
+$(TYPEDSIGNATURES)
+
+Shallow copy of a [`EnvironmentalLink`](@ref)
+"""
+Base.copy(m::EnvironmentalLink) = EnvironmentalLink(;
+    reaction_id = m.reaction_id,
+    metabolite_id = m.metabolite_id,
+    lower_bound = m.lower_bound,
+    upper_bound = m.upper_bound,
+)
+
+"""
+$(TYPEDSIGNATURES)
+
+A helper function that creates an exchange/environmental variable linking matrix
+for community member `m`.
+"""
+function env_ex_member_matrix(
+    m::CommunityMember,
+    env_mets::Vector{String},
+    env_rxns::Vector{String},
+)
+    mat = spzeros(length(env_mets), n_variables(m.model))
+    for (midx, mid) in enumerate(env_mets)
+        mid in metabolites(m.model) || continue # does not have this exchange
+        ridx = first(indexin([env_rxns[midx]], variables(m.model))) #  find index of exchange reaction in member
+        mat[midx, ridx] = 1.0
+    end
+    return mat
+end
+
+"""
+$(TYPEDSIGNATURES)
+
+A helper function that creates the entire exchange/environmental variable
+linking matrix for a community model.
+"""
+function env_ex_matrix(cm)
+    env_mets = [envlink.metabolite_id for envlink in cm.environmental_links]
+    env_rxns = [envlink.reaction_id for envlink in cm.environmental_links]
+
+    hcat(
+        [
+            env_ex_member_matrix(m, env_mets, env_rxns) .* a for
+            (m, a) in zip(cm.members, cm.abundances)
+        ]...,
+    )
+end
+
+"""
+$(TYPEDSIGNATURES)
+
+Variant of [`env_ex_matrix`](@ref) that takes an explicit abundance matrix (used
+in solver modifications.)
+"""
+function env_ex_matrix(cm, abundances)
+    env_mets = [envlink.metabolite_id for envlink in cm.environmental_links]
+    env_rxns = [envlink.reaction_id for envlink in cm.environmental_links]
+
+    hcat(
+        [
+            env_ex_member_matrix(m, env_mets, env_rxns) .* a for
+            (m, a) in zip(cm.members, abundances)
+        ]...,
+    )
+end
+
+"""
+$(TYPEDSIGNATURES)
+
+A helper function to find the index of the appropriate model. Assumes each `id`
+is delimited by `#` that separates the model ID prefix and the original id.
+"""
+function access_community_member(
+    cm::CommunityModel,
+    id::String,
+    accessor::Function;
+    default = nothing,
+)
+    id_split = split(id, "#")
+    idx = findfirst(startswith(first(id_split)), m.id for m in cm.members)
+    isnothing(idx) && return default # can only access inside community member
+    accessor(cm.members[idx].model, string(last(id_split)))
+end
+
+
+"""
+$(TYPEDSIGNATURES)
+
+A helper function to add the id of the community member as a prefix to some string.
+"""
+add_community_prefix(m::CommunityMember, str::String; delim = "#") = m.id * delim * str
```

Comment:
```suggestion
add_community_prefix(m::CommunityMember, str::String; delim = "/") = "$(m.id)$delim$str"
```

Reply:
the "#" or "/" should be a constant btw

Reply:
we use `#` for the enzyme constrained, why change it to `\`?

Reply:
I could make the delimiter a constant, but would have to change that for all the delimiters. I will add this to our todo list for 2.0 and do it later